src/vmm/src/builder.rs

```diff
@@ -510,21 +520,63 @@ pub fn build_microvm_from_snapshot(
-        // No TSC freq in snapshot means we have to fail-fast.
-        let state_tsc = microvm_state.vcpu_states[0]
-            .tsc_khz
-            .ok_or(BuildMicrovmFromSnapshotError::TscFrequencyNotPresent)?;
-
-        // Scale the TSC frequency for all VCPUs, if needed.
-        if vcpus[0].kvm_vcpu.is_tsc_scaling_required(state_tsc)? {
-            for vcpu in &vcpus {
-                vcpu.kvm_vcpu.set_tsc_khz(state_tsc)?;
+    {
+        // Convert CPUID to better format
+        let snapshot_cpuid = {
+            let raw_snapshot_cpuid =
+                cpuid::RawCpuid::from(microvm_state.vcpu_states[0].cpuid.clone());
+
+            // TODO Prevent `microvm_state.vcpu_states[0].cpuid` containing duplicate and invalid
+            // leaves. Do not save snapshots with invalid leaves.
+            //
+            // We need to do this as `microvm_state.vcpu_states[0].cpuid` contains many invalid
+            // leaves if we do not remove them they will overwrite valid leaves (e.g. a trailing
+            // invalid leaf marked as `function: 0, index: 0, eax: 0, ebx: 0, ecx: 0, edx: 0` would
+            // overwrite the manufactuer id). Ideally it should not contain these leaves as this is
+            // wasting memory.
+            //
+            // We cannot remove leaves with 0 in all registers as neither AMD nor Intel note that 0
+            // in all registers means a leaf is invalid. Thus we go off the arbitrary ordering of
+            // `microvm_state.vcpu_states[0].cpuid` and only retain the earliest leaf for which
+            // there are duplicates.
+            //
+            // On the case of support, Intel comes the closest to saying something useful with:
+            // > If a value entered for CPUID.EAX is less than or equal to the maximum input value
+            // > and the leaf is not supported on that processor then 0 is returned in all the
+            // > registers.
+            // Importantly, this allows us to infer a leaf will contain 0 in all registers if we
+            // know a leaf is unsupported. It does not allow us to infer a leaf is unsupported if we
+            // know it contains 0 in all registers. The case that a leaf is supported and contains 0
+            // in all registers is allowed. So this is not really useful.
+            let mut existing = std::collections::HashSet::new();
+            let trimmed_raw_snapshot_cpuid = raw_snapshot_cpuid
+                .iter()
+                .cloned()
+                .filter(|entry|
+                // If already present this returns false, filtering out this element
+                existing.insert((entry.function,entry.index)))
+                .collect::<cpuid::RawCpuid>();
+
+            cpuid::Cpuid::try_from(trimmed_raw_snapshot_cpuid)
+                .map_err(BuildMicrovmFromSnapshotError::CpuidNewError)?
+        };
+
+        // Get KVM supported CPUID
+        let supported_cpuid = cpuid::Cpuid::kvm_get_supported_cpuid()?;
```

Comment:
Ok, we need to think very carefully about this. We want to pay extra attention to any syscalls we do at snapshot time (the target for restoring from snapshot is < 5ms) so any additional syscall we call could end up being very expensive. 

Additionally, This change seems more to be from an area of how do we ensure saved snapshot is compatible with current host so I suggest to drop this here and open this in a separate PR.

Reply:
How much time would this need to add to snapshot restore time to result in the CI failure?

Reply:
Moved to https://github.com/firecracker-microvm/firecracker/pull/3360.